Created Book class and methods to manage book checkout

euloong opened the pull request:
- Created Book class to manage checkedOut status

In MainMenu:
- Created showBookList method to select checkedIn books and format how they are displayed
- Updated books ArrayList to be list of Book objects
- Created manageBookList method to manage user checking books out
- Created countCheckedOutBooks method

To do: fix bug in manageBookList for the count of checkedOutBooks, fix 3 tests failing due to user input loop

andeemarks commented on src/com/twu/biblioteca/Book.java:
```diff
@@ -0,0 +1,35 @@
+package com.twu.biblioteca;
+
+public class Book {
+    private final String title;
+    private final String author;
+    private final int yearPublished;
```
Nice - final attributes mean the Book objects will be mostly immutable and much easier to reason about.

andeemarks commented on src/com/twu/biblioteca/Book.java:
```diff
@@ -0,0 +1,35 @@
+        return checkedOut;
+   }
+
+   public void setCheckedOut() {
```
Is there another name we could use that also describes what is happening here?  setXXX is the convention for a setter method for attribute XXX, but it is  also typical to pass the value that XXX is being set to.

euloong replied:
public void setCheckedOut(Boolean checkedOut){
  this.checkedOut = true;
}

or should it be:

private boolean available; (or make it notAvailable to match my current intention)

public boolean isAvailable(){
return this.available;
}

public void setAvailable(boolean available){
  this.available = available;
}

but would the default state be false?

andeemarks replied:
I would land on just one term (i.e., either "checked out" or "available") and ideally the one that corresponds to the terminology used in the requirements/stories.  Unfortunately, the stories mention both terms, but it seems like "availability" is the status of the book and "checking out" is the verb applied to change that status.

Depending on how far down this thought experiment you want to go, you could consider whether keeping them independent makes sense.  In this case, the attribute would be "available" and it would have it's own set of getter/setter __and__ you would _also_ provide a `Book#checkOut()` method that checks whether the book is available.  _If there are any future requirements_ that talk of other events which result in a book being unavailable (e.g., being repaired), then this might make sense.

There's no right/wrong answers in any of these options, just degrees of simplicity of implementation, simplicity of readability and natural fit with the domain language.  Rarely can you optimise perfectly across all of these aspects simultaneously.

andeemarks commented on src/com/twu/biblioteca/Book.java:
```diff
@@ -0,0 +1,35 @@
+    private final int yearPublished;
+    private boolean checkedOut;
+
+    public Book(String title, String author, int yearPublished, boolean checkedOut) {
```
Does IntelliJ give you any messages about the `checkedOut` parameter here?

euloong replied:
It says that it is never used. I'm not sure why it's saying that.

andeemarks replied:
I'm on the side of IntelliJ here :-)  If you look closely at how you are using the first three params compared to the last one, it might become clearer.  

Because of the static nature of the Java language (and the maturity of IntelliJ), paying close attention to all those IntelliJ warnings is generally a very good idea.  They are often soft pointers to likely bugs.

euloong replied:
I'm not following the convention. It should be:
this.checkedOut = checkedOut;

andeemarks commented on src/com/twu/biblioteca/MainMenu.java:
```diff
@@ -32,36 +34,71 @@ public void showGoodbyeMessage() {
+
+    public int countCheckedOutBooks() {
+        int count = 0;
+        for (Book b : this.books){
```
Could you use this type of loop in other places in this class where you are iterating over an array?  I prefer this version to the classic `for ( ; ; )` format because it's smaller and doesn't require the explicit creation of the `i` index.

euloong replied:
Yes, I believe it's the for-each loop notation and I can apply it to the showOptions method. 

But it has limitations, namely it can't process two decision making statements at the same time.

I have a nested if statement in the for loop in showBookList and therefore the for-each loop would not be appropriate here.

andeemarks commented on src/com/twu/biblioteca/MainMenu.java:
```diff
@@ -32,36 +34,71 @@ public void showGoodbyeMessage() {
+
+    public void manageBookList() {
+        int count = countCheckedOutBooks();
+      if ((this.books.size() - count) == 0) {
```
if `x - y = 0`, then what is the relationship between `x` and `y`?  I ask because there's a clearer way to express this condition.

euloong replied:
Ahh I'm so silly. It's so basic.

x = y

I should be asking if this.books.size() == count

andeemarks commented on src/com/twu/biblioteca/MainMenu.java:
```diff
@@ -9,16 +9,18 @@
     private final String welcome = "Welcome to Biblioteca. Your one-stop-shop for great book titles in Bangalore!";
     private final String invalidMessage = "Please select a valid option!";
     private final String goodBye = "Goodbye!";
+    private final String checkOutBookReference = "Enter the number of the book you want to checkout:";
```
It's your choice  to pull these strings out into constants or not, but there is a heuristic call the [Rule of 3s](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) which people often use to help them work out when the tipping point has been reached for these types of things.

euloong replied:
So maybe keep them in their own separate methods? That seems to be where intelliJ is trying to steer me.

andeemarks replied:
If they are only used in a single method, then declaring them there (as a local variable) or just inlining the strings is probably a better option as the reader of your code doesn't need to scan outside of the method to understand what `checkOutBookReference` means.

euloong replied:
Ok thanks, saves the reader from having to jump up and down in the code.

andeemarks commented on src/com/twu/biblioteca/Book.java:
```diff
@@ -0,0 +1,35 @@
+      return yearPublished;
+   }
+
+   public boolean getCheckedOut() {
```
From the Java spec around property getters:
```
8.3.2 Boolean properties

In addition, for boolean properties, we allow a getter method to match the pattern:

public boolean is<PropertyName>();

This “is<PropertyName>” method may be provided instead of a “get<PropertyName>” method, or it may be provided in addition to a “get<PropertyName>” method. 
```

andeemarks commented on src/com/twu/biblioteca/MainMenu.java:
```diff
@@ -32,36 +34,71 @@ public void showGoodbyeMessage() {
+                int count = countCheckedOutBooks();
+                System.out.println(String.format(
+                        "%d. %s | %s | %d",
+                        (i + 1) - count, book.getTitle(), book.getAuthor(), book.getYearPublished()));
```
The responsibility for formatting a Book for printing could also be put into the Book class itself.  Have a look at https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#toString() and see if there's anything there that might help with this.

euloong replied:
Yes, I can see how that is cleaner. I'll use it for when I break up the MainMenu class.

andeemarks replied:
Doing so will also allow you to do most of the testing of the format inside the Book test as well.

andeemarks commented on src/com/twu/biblioteca/MainMenu.java:
```diff
@@ -32,36 +34,71 @@ public void showGoodbyeMessage() {
+        }
+    }
+
+    public int countCheckedOutBooks() {
+        int count = 0;
+        for (Book b : this.books){
+            if (b.getCheckedOut()) count++;
+        }
+        return count;
+    }
```
Java also (now) has a bunch of methods that operate on collections to support this sort of processing without having to write an explicit `for` loop.  Many of these methods have similar names to JS and provide a far more functional way to write collection processing code.  If you're feeling confident with your Java skills, it's a nice refactoring to look at.

euloong replied:
So I could use an iterator?

andeemarks replied:
Yep!  Java streams/iterators are now at the point where you can replace almost all explicit iterations with an appropriate use of `map`, `reduce`, `collect`, etc.  The syntax is not as clean as JS, but it's not too bad either.

Alternatively, you could make a good argument for the explicit looping being more easy to read, but - as an industry - I think the average developer is getting better at pulling these tools into their standard toolkit, so I encourage their use.

euloong commented:
Hi Andy, 
I'll be able to work on your comments tomorrow. 

Had some difficulties in getting my menus and tests to work correctly so took the last couple of days to fix them. Agree that I need to work on refactoring the responsibilities and naming conventions.

- Fixed bug where updating the displayed index reference was not working in certain combinations.
(It's still not perfect e.g. if book at index 0 is checked out then remaining books are displayed as 2. a book, 3. some book, instead of 1., 2., 3. etc. Tried created a new ArrayList and looping the checked in books into that but couldn't get the loop to print correctly. Might need to have another crack at it another time)

- Fixed the failing tests.
   There was a NoSuchElementException error if the scanner didn't receive an input.
   I had to put a condition in the manageOptions method to break the while loop if no token was provided to the optionsScanner from the test input. (I need to read up on switch with case and break for when I refactor)
